File: Design/1381. Design a Stack With Increment Operation/1381. Design a Stack With Increment Operation.java
class CustomStack {
    // lazy load

    int[] stackarray;
    int[] lazy;
    int pos;
    int size;
    public CustomStack(int maxSize) {
        stackarray = new int[maxSize];
        lazy = new int[1001];
        pos = 0;
        size = maxSize;
    }
    
    public void push(int x) {
        if (isFull()) return;
        stackarray[pos++] = x;
    }
    
    public int pop() {
        if (isEmpty()) return -1;
        if (lazy[pos-1] != 0) {
            int tmp = lazy[pos-1];
            if (pos-2 >= 0) lazy[pos-2] += tmp;
            lazy[pos-1] = 0;
            return stackarray[--pos]+tmp;
        } else {
            return stackarray[--pos];
        }
    }
    
    public void increment(int k, int val) {
        if (isEmpty()) return;
        if (k > pos) {
            lazy[pos-1] += val;
        } else {
            lazy[k-1] += val;
        }
        
    }
    private boolean isEmpty() { return pos == 0; }
    private boolean isFull() { return pos == size; }
}

File: Design/1381. Design a Stack With Increment Operation/1381. Design a Stack With Increment Operation.cpp
class CustomStack {
    vector<int>nums;
    vector<int>offset;
    int maxSize;
    int count = 0;
    int diff = 0;
public:
    CustomStack(int maxSize) {
        this->maxSize = maxSize;
        nums.resize(maxSize, 0);
        offset.resize(maxSize, 0);
    }
    
    void push(int x) {
        if (count == maxSize) return;
        
        if (count > 0) offset[count-1] += diff;
        
        diff = 0;
        nums[count] = x;
        offset[count] = 0;
        count += 1;
    }
    
    int pop() {
        if (count == 0) return -1;
        
        diff += offset[count-1];
        //cout<<diff<<endl;
        return nums[--count] + diff;
    }
    
    void increment(int k, int val) {
        if (count == 0) return;
        offset[min(k-1, count-1)] += val;
    }
};

//       xxxxx y
//offset     3
// diff: 3 -> 0
